File: ch4/traversal/cyclecheck.py
import sys
from enum import Enum

class flag(Enum):
  UNVISITED = -1
  EXPLORED = -2
  VISITED = -3

AL = []
dfs_num = []
dfs_parent = []

def cycleCheck(u):
  global AL
  global dfs_num
  global dfs_parent

  dfs_num[u] = flag.EXPLORED.value
  for v, w in AL[u]:
    if dfs_num[v] == flag.UNVISITED.value:
      dfs_parent[v] = u
      cycleCheck(v)
    elif dfs_num[v] == flag.EXPLORED.value:
      if v == dfs_parent[u]:
        printf(' Bidirectional Edge (%d, %d)-(%d, %d)' % (u, v, v, u))
      else:
        print('Back Edge (%d, %d) (Cycle)' % (u, v))
  dfs_num[u] = flag.VISITED.value


def main():
  global AL
  global dfs_num
  global dfs_parent

  fp = open('scc_in.txt', 'r')

  V = int(fp.readline().strip())
  AL = [[] for _ in range(V)]
  for u in range(V):
    tkn = list(map(int, fp.readline().strip().split()))
    k = tkn[0]
    for i in range(k):
      v, w = [tkn[2*i+1], tkn[2*i+2]]
      AL[u].append((v, w))

  print('Graph Edges Property Check')
  dfs_num = [flag.UNVISITED.value] * V
  dfs_parent = [-1] * V
  for u in range(V):
    if dfs_num[u] == flag.UNVISITED.value:
      cycleCheck(u)


main()

File: ch4/traversal/cyclecheck.ml
open Scanf
open Printf

type graph = int list array (* Adjacency List representation *)

type vertex_state = White | Gray | Black

let graph_check (adj : graph) =
  let n = Array.length adj in
  let color = Array.make n White in
  let parent = Array.make n (-1) in
  let bidirectionals = ref [] in
  let backs = ref [] in
  let forward_cross = ref [] in
  let rec visit u =
    color.(u) <- Gray;
    adj.(u) |> List.iter (fun v ->
      match color.(v) with
      | White -> parent.(v) <- u; visit v
      | Gray ->
        if v = parent.(u) then bidirectionals := (u, v) :: !bidirectionals
        else backs := (u, v) :: !backs
      | Black -> forward_cross := (u, v) :: !forward_cross
    );
    color.(u) <- Black
  in
  List.init n (fun x -> x) |> List.iter (fun u -> if color.(u) = White then visit u);
  !bidirectionals, !backs, !forward_cross

let () =
  let ic = Scanning.open_in "scc_in.txt" in
  let sc = object method nf = bscanf ic end in
  let n = sc#nf " %d " (fun x -> x) in
  let adj = Array.init n (fun _ ->
    sc#nf " %d " (fun degree ->
      List.init degree (fun _ -> sc#nf " %d %d " (fun v w -> v))) (* Ignoring weight *)
  ) in
  printf "Graph Edges Property Check\n";
  let bidirectionals, backs, forward_cross = graph_check adj in
  bidirectionals |> List.iter (fun (u, v) ->
    printf " Bidirectional (%d, %d) - (%d, %d)\n" u v v u);
  backs |> List.iter (fun (u, v) ->
    printf " Back Edge (%d, %d) (Cycle)\n" u v
  );
  forward_cross |> List.iter (fun (u, v) ->
    printf " Forward/Cross Edge (%d, %d)\n" u v
  );
